File: 0987-vertical-order-traversal-of-a-binary-tree/0987-vertical-order-traversal-of-a-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
map<int, vector<pair<int, int>>> mp;  // col → list of (row, val)

    void dfs(TreeNode* root, int row, int col) {
        if (!root) return;

        mp[col].emplace_back(row, root->val);

        dfs(root->left, row + 1, col - 1);  // move left
        dfs(root->right, row + 1, col + 1); // move right
    }
    vector<vector<int>> verticalTraversal(TreeNode* root) {
        dfs(root, 0, 0);

        vector<vector<int>> res;

        for (auto& [col, nodes] : mp) {
            // Sort by row, then by value if rows are the same
            sort(nodes.begin(), nodes.end());
            vector<int> level;
            for (auto& [row, val] : nodes) {
                level.push_back(val);
            }
            res.push_back(level);
        }

        return res;
    }
};

File: 0987-vertical-order-traversal-of-a-binary-tree/0987-vertical-order-traversal-of-a-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    TreeMap<Integer,TreeMap<Integer,PriorityQueue<Integer>>> q=new TreeMap<>();

    public void dfs(TreeNode root,int x,int y){
        if(root==null){
            return;
        }

        q.putIfAbsent(x,new TreeMap<>());
        q.get(x).putIfAbsent(y,new PriorityQueue<>());
        q.get(x).get(y).offer(root.val);

        dfs(root.left,x-1,y+1);
        dfs(root.right,x+1,y+1);
    }
    public List<List<Integer>> verticalTraversal(TreeNode root) {
        dfs(root,0,0);

        List<List<Integer>> ans=new ArrayList<>();

        for(TreeMap<Integer,PriorityQueue<Integer>> pq:q.values()){
            List<Integer> li=new ArrayList<>();

            for(PriorityQueue<Integer> node: pq.values()){
                while(!node.isEmpty()){
                    li.add(node.poll());
                }
            }
            ans.add(li);
        }
        return ans;
    }
}